tests/core/util/test_mock_web_server.py

```diff
@@ -0,0 +1,241 @@
+        if _readable > 0:
+            _request.payload = self.rfile.read(_readable)
+        return _request
+
+    def _handle_everything(self):
+        _request = self._read_everything()
+        _response = self.server.mock_api_server.dequeue_response(_request)
+        if _response is None:
+            logging.error(
+                f"failed to find a response for {_request.method} {_request.path}"
+            )
+            raise AssertionError("No available response!")
+        self._send_everything(_response)
+
+    def do_GET(self):
+        logging.info("GET")
+        self._handle_everything()
+
+    def do_POST(self):
+        logging.info("POST")
+        self._handle_everything()
+
+    def do_PUT(self):
+        logging.info("PUT")
+        self._handle_everything()
+
+    def version_string(self) -> str:
+        return ""
+
+    def date_time_string(self, timestamp: Optional[int] = 0) -> str:
+        return "Sat, 1 January 2000 00:00:00 UTC"
+
+
+class MockAPIInternalServer(HTTPServer):
+    mock_api_server: "MockAPIServer"
+
+    def __init__(self, server_address: Tuple[str, int], bind_and_activate: bool):
+        super(MockAPIInternalServer, self).__init__(
+            server_address, MockAPIServerRequestHandler, bind_and_activate
+        )
+        self.allow_reuse_address = True
+
+
+class MockAPIServer:
+    """Embedded web server."""
+
+    _address: str
+    _port: int
+    _server: HTTPServer
+    _server_thread: threading.Thread
+    _responses: Dict[str, Dict[str, List[MockAPIServerResponse]]]
+    _requests: List[MockAPIServerRequest]
+
+    def __init__(self, address: str, port: int):
+        self._address = address
+        self._port = port
+        self._server = MockAPIInternalServer(
+            (self._address, self._port), bind_and_activate=True
+        )
+        self._server.mock_api_server = self
+        self._server_thread = threading.Thread(target=self._server.serve_forever)
+        self._responses = {}
+        self._requests = []
+
+    def start(self) -> None:
+        self._server_thread.start()
+
+    def stop(self) -> None:
+        self._server.shutdown()
+        self._server.server_close()
+        self._server_thread.join(timeout=10)
+
+    def enqueue_response(
+        self, request_method: str, request_path: str, response: MockAPIServerResponse
+    ):
+        _by_method = self._responses.get(request_method) or {}
+        _by_path = _by_method.get(request_path) or []
+        _by_path.append(response)
+        _by_method[request_path] = _by_path
+        self._responses[request_method] = _by_method
+
+    def dequeue_response(
+        self, request: MockAPIServerRequest
+    ) -> Optional[MockAPIServerResponse]:
+        self._requests.append(request)
+        _by_method = self._responses.get(request.method) or {}
+        _by_path = _by_method.get(request.path) or []
+        if len(_by_path) > 0:
+            return _by_path.pop(0)
+        return None
```

Comment:
Best to raise an error so the test fails predictably if there was no response enqueued for this request

Reply:
That actually happens here:

https://github.com/ThePalaceProject/circulation/pull/257/files/aad54d9e8367b0ee2c2d3306206eb418c118ac79#diff-9712f5dfb46342a5dd008eed3f73df687da72fce71a2f897e3d814cb77000e13R77

Reply:
It'll cause the connection to be terminated, so unless a test is specifically checking for connection failures, it won't accidentally succeed because the server happened to yield an HTTP status code that the test was looking for.